src/journal/journald-socket.c

```diff
@@ -0,0 +1,110 @@
+
+        assert(s);
+
+        if (s->forward_socket_fd >= 0) {
+                log_debug("Forward socket already exists, refusing to open a VM forward socket.");
+                return;
+        }
+
+        socket_fd = socket(AF_VSOCK, SOCK_STREAM|SOCK_CLOEXEC, 0);
+        if (socket_fd < 0) {
+                log_debug_errno(errno, "Failed to create AF_VSOCK socket, ignoring: %m");
+                return;
+        }
+
+        if (connect(socket_fd, &addr->sockaddr.sa, addr->size) < 0) {
+                log_debug_errno(errno, "Failed to connect to vsock address for forwarding, ignoring: %m");
+                return;
+        }
+
+        s->forward_socket_fd = TAKE_FD(socket_fd);
+        log_debug("Successfully connected to vsock for forwarding");
+}
+
+void server_detect_unix_forward_socket(Server *s) {
+        _cleanup_close_ int socket_fd = -EBADF;
+        int r;
+
+        assert(s);
+
+        if (s->forward_socket_fd >= 0) {
+                log_debug("Forward socket already connected, not detecting unix socket.");
+                return;
+        }
+
+        socket_fd = socket(AF_UNIX, SOCK_STREAM|SOCK_CLOEXEC, 0);
+        if (socket_fd < 0) {
+                log_debug_errno(errno, "Failed to create AF_UNIX socket, ignoring: %m");
+                return;
+        }
+
+        r = connect_unix_path(socket_fd, AT_FDCWD, "/run/host/journal/socket");
+        if (r < 0) {
+                log_debug_errno(r, "Failed to connect to /run/systemd/journal/stdout, ignoring: %m");
+                return;
+        }
+
+        s->forward_socket_fd = TAKE_FD(socket_fd);
+        log_debug("Successfully connected to unix socket for forwarding");
+}
+
+void server_forward_socket(
+                Server *s,
+                const struct iovec *iovec,
+                size_t n_iovec,
+                int priority) {
+        _cleanup_free_ struct iovec *iov_alloc = NULL;
+        struct iovec *iov = NULL;
+
+        assert(s);
+        assert(iovec);
+        assert(n_iovec > 0);
+
+        if (LOG_PRI(priority) > s->max_level_socket)
+                return;
+
+        /* we need a newline after each one + two on the final one */
+        size_t n = n_iovec * 2;
+        if (n < ALLOCA_MAX / sizeof(struct iovec) / 2)
+                iov = newa(struct iovec, n);
+        else {
+                iov_alloc = new(struct iovec, n);
+                if (!iov_alloc) {
+                        log_oom();
+                        return;
+                }
+
+                iov = iov_alloc;
+        }
+
+        char lf[] = "\n", lflf[] = "\n\n";
+        for (size_t i = 0; i < n_iovec; i++) {
+                iov[i * 2] = iovec[i];
+                iov[i * 2 + 1] = (struct iovec) {
+                        .iov_base = lf,
+                        .iov_len = 1
+                };
+        }
+        iov[n - 1] = (struct iovec) {
+                .iov_base = lflf,
+                .iov_len = 2,
+        };
```

Comment:
hmm, which protocol is this?

note that journal records can be binary, newlines alone are not good separators hence.

THis should probably follow this protocol:

https://systemd.io/JOURNAL_NATIVE_PROTOCOL

That protocol is built around datagrams however. Here we have a stream socket. hence maybe take inspirition from howe DNS-over-TCP relates to DNS-over-UDP: simply write a 64bit integer of the size of the next datagram followed by the datagram payload and so on.

Reply:
I'm fairly sure this should be the journal export format?

Reply:
I like the idea of changing it to the journal native protocol a lot.
One issue I had been confronting is how to actually get the logs to journald from nspawn/vmspawn.
And with the native journal format I think I could just pass all the packets on directly.
Or even in the nspawn case mount that socket inside the container? (assuming this is safe)

Reply:
so using the native proto means that if the messages get too large we'll hit the message size limit and are in theory supposed to send a memfd in that case. But I think this is a non-issue here, we don't want such large objects anyway via these transports, hence drop and write a short log message if we can#t forward it.



> One issue I had been confronting is how to actually get the logs to journald from nspawn/vmspawn. And with the native journal format I think I could just pass all the packets on directly.

So you could do that, but do note that journald enforces restrictions on the fields you can send. i.e. all fields starting with an underscore (i.e. _PID=, _UID=, …) are controlled by journald, not by the sender, i.e. it always writes them itself so that you can trust them. Thus, these structured fields would be lost as you proxy the messages. Hence maybe prefix these with something (i.e. PAYLOAD_UID= instead of _UID= and so on, you get the idea)

> Or even in the nspawn case mount that socket inside the container? (assuming this is safe)

the _UID= fields would then be subject to uidmapping if that's used, and that would be very confusing...

Reply:
Note that it would be rather useful if we could use older versions of systemd-journal-remote to read whatever is sent by journald from the VM and I don't think systemd-journal-remote knows how to read the native format.

Reply:
The format now sent is Export format. In order that systemd-journal-remote can read it the extra metadata field __REALTIME_TIMESTAMP= is synthesised as the final field from `now(CLOCK_REALTIME)`.

I've tested this locally and systemd-journal-remote will happily accept all of the content sent and save it to a journal file!